### ISSUE TITLE
No error for array index passed by ref

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#13062]
No error reported for `a[ref i]` below:
```
class C
{
    static object F(object[] a, int i)
    {
        return a[ref i];
    }
}
```

**Comment by gafter** [comment]
This is apparently an intentional bug maintained for backward compatibility. From `LanguageParser.cs`, we read

```cs
                // According to Language Specification, section 7.6.7 Element access
                //      The argument-list of an element-access is not allowed to contain ref or out arguments.
                // However, due to backward compatibility, compiler overlooks this restriction during parsing
                // and even ignores out/ref modifiers in element access during binding.
                //
                // We will enforce that language rule for out variables declarations at the parser level. 
```

**Comment by gafter** [comment]
I will add a note to the compiler specification.

**Comment by cston** [comment]
Pre-Roslyn, the C# compiler reported:
```
error CS0029: Cannot implicitly convert type 'ref int' to 'int'
```

**Comment by gafter** [comment]
I've submitted a PR to diagnose this in strict mode.

**Comment by gafter** [comment]
> Pre-Roslyn, the C# compiler reported:
```
error CS0029: Cannot implicitly convert type 'ref int' to 'int'
```

I think at one time (when we first had a functional Roslyn compiler) the native compiler did not diagnose this. That was why Roslyn intentionally let this through for compatibility. I'll talk to the compat council about whether we can begin diagnosing this now.

**Comment by VSadov** [comment]
I think we should consider just giving error here. The code is clearly bogus. IMO, the chances of hitting compat issues here are extremely low.

**Comment by gafter** [comment]
@VSadov The reason the code allows this is because we did run into a compat issue.

**Comment by VSadov** [comment]
@gafter  did we observe the compat issue in a testcase whose only purpose was to perpetuate this behavior or there was some actual code that does this while not being a testcase?

**Comment by gafter** [comment]
OK, making it an error, with approval of the @dotnet/roslyn-compat-council
